### ISSUE TITLE
Inavlid CSRF Token using Single page application(Angular) in express js

### ISSUE THREAD

**Issue by Ahsanbaloch** [expressjs/csurf#227]
can you help me to sort out the issue. I have tried your solution but still getting error of { "message": "invalid csrf token" }

Here is the code of serve.js of express
```
require('rootpath')();
const express = require('express');
const app = express();
const helmet = require("helmet");
const bodyParser = require('body-parser');
const csrf = require('csurf')
const cookieParser = require('cookie-parser')

const cors = require('cors');
const errorHandler = require('_middleware/error-handler');

app.use(helmet({ dnsPrefetchControl: { allow: true }}));
app.use(bodyParser.urlencoded({ extended: false }));
app.use(bodyParser.json());

app.use(cookieParser());

app.use(csrf({cookie: {key: 'XSRF-TOKEN'}}))

// app.use('/',  function (req, res) {
//     res.cookie('XSRF-TOKEN', req.csrfToken());
//     // res.render('server')
//   })


app.use(cors({ origin: (origin, callback) => callback(null, true), credentials: true }));



// api routes
app.use('/accounts', require('./accounts/accounts.controller'));

// swagger docs route
app.use('/api-docs', require('_helpers/swagger'));

// global error handler
app.use(errorHandler);

// start server
const port = process.env.NODE_ENV === 'production' ? (process.env.PORT || 80) : 4000;
app.listen(port, () => console.log('Server listening on port ' + port));
```
and the angular code of module.ts is
```
HttpClientXsrfModule.withOptions({
          cookieName: 'XSRF-TOKEN',
          headerName: 'X-XSRF-TOKEN'
        })
```
Currently I am checking output on postman and also on angular but still get no expected output.

**Comment by avizuber** [comment]
Was this ever resolved?

**Comment by Ahsanbaloch** [comment]
> Was this ever resolved?

yeah.
I have tried another NPM and it worked properly. Here it is.
https://www.npmjs.com/package/express-csrf-double-submit-cookie

**Comment by dougwilson** [comment]
The issue with your above code is that you are setting the cookie `XSRF-TOKEN` to the csrf secret, not the csrf token, which is the cookie you need to tell angular to read. Your server code never sends the csrf token anywhere, as you commented that part out. You need to uncomment your `res.cookie` middleware and change the name of the cookie for this middleware not to be the same for it to work.

**Comment by Ahsanbaloch** [comment]
> The issue with your above code is that you are setting the cookie `XSRF-TOKEN` to the csrf secret, not the csrf token, which is the cookie you need to tell angular to read. Your server code never sends the csrf token anywhere, as you commented that part out. You need to uncomment your `res.cookie` middleware and change the name of the cookie for this middleware not to be the same for it to work.

I have done but still getting error of invalid csrf token

**Comment by dougwilson** [comment]
Hi @Ahsanbaloch sorry you are still having problems. In order to help, we need you to provide the updated code and all other parts so we can run it just like you are. I used your code above with the modifications I described, plus of course I had to create my own Angular app, etc. since you didn't provide it just to test. So there are likely a lot of differences between our reproduction cases that could be the culprit. Providing enough code to run and reproduce the issue is the best way for us to assist you fully.
